### ISSUE TITLE
model witness values

### ISSUE THREAD

**Issue by LeventErkok** [cvc5/cvc5#8898]
@ajreynol 

Looks like `--model-witness-value ` flag was removed in this commit: 7de44af3d352fe9f622945f8a61b2b1b4e6d8b49

In our upstream tool, we use this option, as discussed at length in #3300 

The commit says "We now allow partial specifications. Recording approximations is arithmetic specific, and if necessary could be added as a post-processing analysis e.g. during check-model."

Is there an example on how to do this from the SMTLib interface to regain this capability?

To be honest, this isn't a huge deal; if you no longer want to support this facility, we can drop support for it in our upstream-tool as well. Just want to understand what the current thinking on this is.

**Comment by ajreynol** [comment]
Indeed, this was removed, sorry for the inconvenience this may have caused. 

This feature was removed for the sake of consistency, and since the non-linear solver in cvc5 was refactored to support real algebraic numbers in models. Furthermore, we have had several new cases where a theory wanted a model to be under-specified. In general, we've found the easiest way to allow this is to not use a pre-existing (possibly unevaluated) term the variable is equal to, instead of insisting on a value encapsulated by a witness term.

The cleaner way for us to support the former feature would be to make this independent of the core model construction (which that PR message alludes to), although if we did this, I think it'd be best to redesign this interface altogether, e.g. with a `(get-witness-value)` command, or more generally supporting approximations in models.

If this becomes a priority for you, do let me know and I can work on adding an interface for this.

**Comment by LeventErkok** [comment]
Thanks @ajreynol 

The issue isn't really about getting an approximate value. (Though that'd be nice of course!) The real issue is when we issue a `get-value` command, there's no indication that the value returned by CVC isn't actually accurate. (At least this was the case in CVC4; the classic `x^2=2` problem doesn't terminate with CVC5 for me.) That is, an upstream tool issuing `get-value` on a real-valued variable has no indication that the value printed isn't precise. (This was discussed in depth at #3300)

It'd be really nice to have an indication that  the value `get-value` returns isn't accurate; it might even simply state that: i.e., it can say the value is not rational for it to print it out in the output. This'd trigger the upstream tool to at least recognize the possibility and take action. As it stands, there doesn't seem to be an easy way for the upstream tool to tell if the response gotten for `get-value` is precise or not.

Do you think this might be added as a feature? (i.e., some indication that the value isn't rational and thus not exact; though not necessarily putting the approximate value in the answer.)

**Comment by LeventErkok** [comment]
Thanks @ajreynol 

In CVC4, you could solve problems of the form:

```
(set-option :global-declarations true)
(set-option :produce-models true)
(set-logic ALL)
(declare-fun s0 () Real)
(assert (= (* s0 s0) 2.0))
(check-sat)
(get-value (s0))
```

and get:

```
sat
((s0 (witness ((BOUND_VARIABLE_833 Real)) (and (>= BOUND_VARIABLE_833 (/ (- 1482917) 1048576)) (>= (* (- 1.0) BOUND_VARIABLE_833) (/ 741455 524288))))))
```

When I try the same with cvc5, it simply loops. Are you saying that (since CVC5 models are always precise), CVC5 will never be able to solve problems of this sort?

This is perfectly fine of course. I want to understand if CVC5 looping on the above problem is just a temporary glitch or if that's always going to be the case since the solution is irrational. And if CVC5 will answer such a query, what will the output look like? That's really the crux of the issue here for upstream tools.
